pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/observables/ObservablesTNCudaOperator.hpp

```diff
@@ -129,6 +144,52 @@ template <class TensorNetT> class ObservableTNCudaOperator {
         add_obs_(obsKey, gateMap.getGateData(obs_name, obs_param));
     }
 
+    /**
+     * @brief Create a map of modes to observable meta data.
+     *
+     * @param obs An observableTNCuda object.
+     * @param modes Modes of all observable terms.
+     * @param term_idx Index of the first observable term.
+     * @param term_idy Index of the second observable term.
+     *
+     * @return std::unordered_map<int32_t, std::vector<MetaDataT>> Map of modes
+     * to observable meta data.
+     */
+    auto create_modes_obsname_map_(ObservableTNCuda<TensorNetT> &obs,
+                                   const vector3D<int32_t> &modes,
+                                   const std::size_t term_idx,
+                                   const std::size_t term_idy)
+        -> std::unordered_map<int32_t, std::vector<MetaDataT>> {
+        std::unordered_map<int32_t, std::vector<MetaDataT>> modes_obsname_map;
+
+        auto modes_termx = modes[term_idx];
+        auto modes_termy = modes[term_idy];
+
+        for (std::size_t tensor_idx = 0; tensor_idx < modes_termx.size();
+             tensor_idx++) {
+            PL_ABORT_IF_NOT(modes_termx[tensor_idx].size() == 1,
+                            "Only one-wire observables are "
+                            "supported for cutensornet v24.03");
+
+            modes_obsname_map[modes_termx[tensor_idx][0]] = {
+                obs.getMetaData()[term_idx][tensor_idx]};
+        }
+
+        for (std::size_t tensor_idy = 0; tensor_idy < modes_termy.size();
```

Comment:
How big could be the  vector3D `modes`?

Reply:
It's system dependent. The number of obs terms can be thousands or more for 20+ qubits if I remember correctly.

Reply:
This part of the code is computed in the device or in the host?
Do you think that it is possible to parallelize these loops with OMP?

Reply:
yea, this is a good idea to parallelize it with OMP! We can come back to it later.